blog/2020-02-16-debezium-on-rds.adoc

```diff
@@ -0,0 +1,226 @@
+* Heartbeats were not enabled. Heartbeats are needed to control the https://debezium.io/documentation/reference/1.0/connectors/postgresql.html#wal-disk-space[WAL disk space consumption] in the following cases:
+** There are many updates in a monitored database but only a minuscule amount relates to the monitored table(s) and/or schema(s). We handled this case by enabling heartbeats by setting `heartbeat.interval.ms`.
+** The PostgreSQL instance contains multiple databases where the monitored database is low-traffic in comparison to the others. Since the WAL is shared by all databases in an instance it keeps accumulating data which cannot be removed until Debezium reads it. But since the high traffic databases are not monitored Debezium is not able to communicate to the database that the WAL files can be removed to reclaim disk space. To solve this scenario we triggered "heartbeat" events by periodically updating a single row in a table created in the monitored database using the below query:
+
+[source,sql]
+----
+CREATE TABLE IF NOT EXISTS heartbeat (id SERIAL PRIMARY KEY, ts TIMESTAMP WITH TIME ZONE);
+INSERT INTO heartbeat (id, ts) VALUES (1, NOW()) ON CONFLICT(id) DO UPDATE SET ts=EXCLUDED.ts;
+----
+
+* We got severely reduced throughput on tables with JSONB columns. After debugging we were able to confirm the reason as frequent schema refresh by Debezium due to TOASTed columns not being present in the replication message. This was fixed by changing `schema.refresh.mode` to `columns_diff_exclude_unchanged_toast` and has since been https://debezium.io/documentation/reference/1.0/connectors/postgresql.html#discrepance-between-plugins[documented].
+* We observed frequent EOF errors on the database connection on a few RDS instance sizes. We are still not sure of the cause but initial investigations point to the issue happening only on instances that have PgBouncer attached (even if not connecting through PgBouncer) or instances with smaller sizes (AWS t2/t3 series).
+* We used a single Debezium connector per PostgreSQL database (instead of per host). The main reasons for not running a single connector per PostgreSQL instance were regarding workload isolation. Any team performing bulk data updates or deletions or unplanned schema migrations will only impact their own Debezium connector instead of the entire PostgreSQL instance since Debezium filters events at its end according to the database and/or schema whitelist and blacklist configured. We are trying to identify possible issues in this configuration but haven't found any yet. Although multiple replication slots on a single database does add overhead, we are able to run fine with around 6 to 10 slots per database host without any noticeable performance impact.
```

Comment:
Perhaps "We *initially* used a single..., but later on switched to a scheme of (a connector per use case|team etc.). The main reasons were ...

Reply:
"*do* add overhead"